### ISSUE TITLE
Throwing Players On Touch

### ISSUE THREAD

**Issue by whipowill** [timbergeron/QSS-M#20]
This is kind of a shot in the dark, but let me explain a bug I'm experiencing.  Note that I'm not entirely sure what is causing this, but I've spent a good amount of time trying to track it down and do the detective work.

I'm using QSS-M with Scarlet mod, which includes Copper.

The bug that I've had the entire time I've been building my mod regards close quarters with other players.  If two players collide, even teammates, sometimes it can cause one of the players to careen off at great speed.  What I've noticed in particular is if I'm touching another player, and if I turn as quickly and wildly as I can, I can essentially "throw" that player across the map.

I can't find anywhere in my QuakeC code where I can control this.  Also, I can't test it on vanilla Quake (no mod), bc w/out the bots I don't have another player to "throw".  My suspicion is this might be from ``ClipVelocity()`` in your ``sv_phys.c`` file.

What do you think?  Any help would be appreciated, bc this bug is making melee engagements frustrating bc players end up bouncing off each other.

EDIT:  My buddy ChatGPT suggests this may be the issue, but he doesn't always know what he's talking about.

![image](https://github.com/timbergeron/QSS-M/assets/44954184/5eedacb0-96cd-4ee4-8ca2-8721639f9d96)

**Comment by timbergeron** [comment]
Hey there! 👋 Apologies for the delayed response — I'll be keeping a closer eye on GitHub issues from now on. Did you manage to resolve the problem you were experiencing? If not, I'd be happy to send over some dev builds for you to test out and hopefully get it sorted. Let me know!

**Comment by whipowill** [comment]
I never did get it figured out, I thought it was an engine issue.  

I had tried in my mod to make melee attacks sort of lunge the player at the enemy, but that would end up throwing that enemy across the map.  It's been a while since I've tested this issue but if I remember correctly I tested this w/ no mods on at all and it still was an issue in the unmodded game.  So that led me to believe it was an engine issue.

It would be cool to get it fixed, but I'm fearful I may not be able to test your builds bc I am on an older Linux version and the latest QSS-M builds don't work for me.  So I play (daily I might add!) on a version from a few years ago.

**Comment by whipowill** [comment]
It occurs to me that I do have a Windows machine (that I loathe) which I could use to test custom builds.  I'd be happy to do it!

**Comment by timbergeron** [comment]
Cool. Is you using linux tied to https://github.com/timbergeron/QSS-M/issues/17 still? Maybe we fix that first?

**Comment by whipowill** [comment]
Yes, that issue still persists for me.  Tried it just now using the QSS-M from your website:

```
./QSS-M-l64: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by ./QSS-M-l64)
./QSS-M-l64: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by ./QSS-M-l64)
```

This build is dated March 2023.

EDIT: The build I've been using bc it still works for me is dated August of 2022.

**Comment by whipowill** [comment]
I would hate to send you on a wild goose chase.  Maybe the Frikbot mod that I included in my mod has something that causes this.

**Comment by timbergeron** [comment]
Long update :)

Gemini 3 came up with this:

QC Band-Aid

[qc.patch](https://github.com/user-attachments/files/24014024/qc.patch)

Engine Bug Fix:

[scarlet.patch](https://github.com/user-attachments/files/24014022/scarlet.patch)

**Comment by timbergeron** [comment]
Fable 5

[q20_scarlet_fix.patch](https://github.com/user-attachments/files/28888853/q20_scarlet_fix.patch)

## Root cause found — it's QC (Copper/Scarlet + mod features), not the engine

I reproduced this headlessly: dedicated QSS-M server, Scarlet, `weapons 1` (axe only), 8 frikbots brawling on dm4, with an instrumented engine build that logs any QC call that massively increases an entity's velocity, with function-level attribution. Two distinct QC mechanisms produce the "thrown across the map" effect. The engine itself never adds velocity to a touched player — every launch traced back to QC.

### Mechanism 1: telefrag knockback = 400,000 ups (`tdeath_touch` → `T_Damage`)

```
VELSPIKE: tdeath_touch() other (player/Adeney)    127 -> 399725 vel(-207480 -6011 -341608)
VELSPIKE: tdeath_touch() other (player/Framan)    222 -> 400127 vel(-252380 -196449 -240445)
VELSPIKE: tdeath_touch() other (player/Theroulde) 323 -> 399405 vel(-217090 -33842 -333543)
VELSPIKE: tdeath_touch() other (player/Furnival)   55 -> 399612 vel(-254819 174187 -253802)
```

That's exactly `50000 * 8`. The chain:

1. Copper spawns a `teledeath` trigger (`spawn_tdeath`, with `force_retouch = 2`) at **every respawn point and teleporter exit** for ~0.2s — the telefrag mechanism.
2. Anyone overlapping that spot gets `T_Damage(other, tdeath, tdeath, 50000)` (`tdeath_touch`, `t_tele.qc`).
3. In `T_Damage` (`combat.qc`) the "figure momentum add" block runs **before** the godmode/invuln/teamplay checks and is uncapped: `targ.velocity += normalize(dir) * 50000 * 8`.
4. If the victim is protected (pent/god/teamplay/undying) they fly across the map **alive**. If they die (usual case), Copper's `ThrowHead` does `self.velocity +=` (vanilla overwrites with `=`), so the corpse/head keeps the 400k velocity — the death camera follows it and, with persistent corpses, everyone watches a body rocket away. The engine clamps it to `sv_maxvelocity` (2000/axis ≈ 3464 ups) the next frame — still trivially "across the map".

Melee scrums make this frequent: constant deaths → teledeaths constantly popping up under the brawl. "Turning quickly while touching" looks like a red herring.

**Fix** (`combat.qc`): skip knockback for teledeath inflictors and cap the rest at quad-rocket level (preserves all normal knockback incl. quad rocket jumps):

```c
	// figure momentum add
	if ( (inflictor != world) && (targ.movetype == MOVETYPE_WALK)
		&& inflictor.classname != "teledeath" && inflictor.classname != "teledeath2" )
	{
		local float kdmg = damage;
		if (kdmg > 480) kdmg = 480;
		dir = targ.origin - BoundsCenter(inflictor);
		dir = normalize(dir);
		targ.velocity = targ.velocity + dir*kdmg*8;
	}
```

### Mechanism 2: corpse football doubles velocity per touch (`KickGibs`, on by default)

With mechanism 1 fixed, the remaining launches all came from `KickGibs` (`scarlet/football.qc`, `scarlet_football` defaults to 1):

```
VELSPIKE: KickGibs() self (player/Quievremont)  46 -> 1633 vel(-362 1500 533)
VELSPIKE: KickGibs() self (player/Puchot)        0 ->  842 vel(-201 736 357)
```

`KickGibs` sets the kicked body to **2× the toucher's velocity**. Two problems:

- Corpses keep `classname == "player"`, so a *flying corpse* re-kicks other corpses — velocity doubles on every corpse-to-corpse touch and cascades (46 → 1633 above is a corpse kicked by another flying body).
- Since `ThrowHead`/corpses are the dead player's own entity, this is again "a player flying across the room" to anyone watching.

**Fix** (`scarlet/football.qc`): require the kicker to be alive and cap the kick:

```c
    // only a living player can kick it
    if (other.classname != "player" || other.health <= 0)
        return;
    ...
    // cap kick speed
    local float spd = vlen(v);
    if (spd > 700)
        v = v * (700 / spd);
```

### Verification

Same setup with the fixed `progs.dat`, cranked to 15 bots so spawn telefrags happen constantly: six telefrag events fired in the first minute (instrumentation still attributes them to `tdeath_touch`), but victims now just gib normally with a ~800 ups pop (`Killed()` clamps health to -99, so gib velocity stays sane) instead of being hurled at 400,000 ups. Corpse kicking still works, capped at 700 ups, and no longer cascades. Over comparable run lengths: unfixed = four 400k launches in ~5 minutes; fixed = zero events above ~1100 ups (all remaining spikes are normal gib tosses and falls). Patch attached.

(Also: please don't use the earlier `scarlet.patch` engine change — clamping the other entity's velocity gain after `SV_Impact` to +100 ups would break rocket knockback / rocket jumps, and the bug isn't in the engine.)
